infer/internal/ende/ende.go

```diff
@@ -188,8 +189,17 @@ func (e *ende) walk(
 			if typ == nil || typ.Kind() != reflect.Struct {
 				return e.walkMap(v, path, elemType, alignTypes)
 			}
-		// This is a scalar value, so we can return it as is.
+		// This is a scalar value, so we can return it as is. The exception is types.AssetOrArchive, which we translate
+		// to a types.Asset or types.Archive, depending on what it contains, to match the SDK's AssetOrArchive type.
 		default:
+			if typ == reflect.TypeOf(types.AssetOrArchive{}) {
+				if v.IsAsset() {
+					v = resource.NewPropertyValue(types.AssetOrArchive{Asset: v.AssetValue()})
+				} else if v.IsArchive() {
+					v = resource.NewPropertyValue(types.AssetOrArchive{Archive: v.ArchiveValue()})
```

Comment:
I'm amazed that this works, but if tests show that it does...

Reply:
Could you elaborate on what you expected?

Reply:
I'm not sure what I expected here. Probably an end run around pulumi/pulumi's mapper.